src/main/java/com/google/neighborgood/servlets/TaskServlet.java

```diff
@@ -93,8 +96,24 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
       }
       out.append("<div class='task-container'>");
       out.append("<div class='task-header'>");
+
+      // Retrieves task's user to get user's nickname
+      String taskOwner = (String) entity.getProperty("Owner");
+      Query userQuery =
+          new Query("UserInfo")
+              .setFilter(new FilterPredicate("userId", FilterOperator.EQUAL, taskOwner));
+      PreparedQuery userResults = datastore.prepare(userQuery);
+      Entity userEntity = userResults.asSingleEntity();
```

Comment:
Thinking about the context of this code, what would you expect to happen when there are lots of tasks in play?

Reply:
Since only 20 tasks are being queried every time the homepage loads and the stress test done by Leonard on retrieving users's info returned results in less or about 10ms, the extra time that it takes to retrieve each task's user's nickname should remain around 0.2s, which I think is reasonable. However, the task query for gathering the most recent tasks might be more time-consuming as there are more and more tasks at play, as it must sort all the tasks before knowing which are the 20 most recent. So  the response time will grow with the task size (I wasn't able to find documentation on what sorting algorithm datastore uses but if it is O(nlogn) as many sorting algorithms are, then response time for sorting tasks would grow linearithmiclly with the total number of tasks). As we aren't expecting many thousands (millions?) ) of tasks (not to mention the datastore limitation of 1000 entitities), I think this is reasonable as well, but let me know if you think this is a faulty assumption!

Reply:
Just curious, where does the limit of 1000 entities come from? I'm not seeing it here: https://cloud.google.com/datastore/docs/concepts/limits.

Having just reviewed #54, I think I can say that the 10ms number comes from using a local testing version of Datastore, rather than what you'll be working with while running in production. Now, it might be better or worse than that in production, but I don't it's necessarily been measured. We can also trust the database to have a reasonable sorting algorithm, so we don't need to be too worried about using the Datastore API to do the ordering for us.

Anyway, for now you're only fetching 20 tasks at a time, sure. Still, you're making an additional round-trip to Datastore for each of those tasks to look up its owner's nickname (which you might have already looked up for previous tasks). I generally try to avoid doing database lookups in a loop, preferring to get more information out of (perhaps more complicated) queries. Can you think of a way to reduce the number of database lookups in this handler?

Reply:
Leonard came across that 1000 entities limitation when searching for the testing documentation, I think he meant this part: "Maximum number of keys allowed for a Lookup operation in the Datastore API". I think this means that if we have more than 1000 entities we can't do `datastore.get(taskKey)`, but we'd still be able to do queries, so it might not be as much of a limitation as we thought.

Ah, that makes sense about the 10ms being just the local testing version. 

To avoid as many queries, specially for tasks from the same owner, I could cache the user nicknames, and check if it's been retrieved already before doing another database lookup. I think that would increase performance, specially if there are multiple tasks by the same user. I could also use projection queries to get just the user's nickname from the UserInfo entities, this would also reduce lookup cost. I'm not sure I got what you meant with avoiding database lookups in a loop and instead getting more info out of queries. I'm not sure it's possible to use one query across multiple entity kinds unless there is an ancestor relationship, but I don't think that applies here since the ancestor in this case would be the user.

Thanks for pushing me to find all these alternative solutions to enhancing the request's performance, I really appreciate it!

Reply:
> I think he meant this part: "Maximum number of keys allowed for a Lookup operation in the Datastore API". I think this means that if we have more than 1000 entities we can't do `datastore.get(taskKey)`,

I interpret this as `datastore.get(taskKey1, taskKey2, taskKey3, ...)` only allows you to look up 1000 tasks at a time *by their keys*. I think you can have more entities than that, but a single Lookup operation (done by `datastore.get`) can only ask for 1000 by their keys explicitly.

And yep, good ideas about reducing the number of database queries. I like the solution you've come up with where you only look them up when needed. In the future, I might consider doing a single project query to get all the user nicknames and put them into a map, but of course that starts to get out of control when you have a lot of users.

Reply:
That makes a lot of sense, it’s good to know we can exceed 1000 entities!

Sounds good! :)

As for the suggestion of doing a project query and storing it in a map - would we need to use something like an http session to store the map so it doesn’t go away after the request is done or how are these things usually handled?

Reply:
You're on the right track - you might eventually set up some kind of external cache that persists across sessions and is available to all replicas of the server. Or, you could store the result with a short timeout ("memoization") in some variable with a broader scope than the doGet function, so that future requests on that server don't have to re-do the lookup for some amount of time. (The user session is typically used for this-user-specific data (e.g., shopping cart items, user preferences, etc.), rather than something like this.)

But even request-specific caching of user lookups is an improvement over doing N user queries for every HTTP request to this handler. I don't see a need to go beyond that in this PR.

Reply:
This makes perfect sense, thanks for clarifying! I might look into this further after going over some of the post-MVP features. It'd be nice to cache these externally if time allows.